src/adapter/src/coord/statement_logging.rs

```diff
@@ -254,7 +254,7 @@ impl Coordinator {
             (StatementLifecycleHistory, statement_lifecycle_updates),
             (SqlText, sql_text_updates),
         ] {
-            if !updates.is_empty() {
+            if !updates.is_empty() && !self.controller.read_only() {
                 self.controller
                     .storage
                     .append_introspection_updates(type_, updates)
```

Comment:
Have you considered moving the `read_only` check into `append_introspection_updates` so callers don't need to remember doing it? I guess the trade-off is that callers might not be aware that their updates won't get appended when the system is in read-only mode, but maybe it's sufficient to have that caveat in the documentation of append-only collections.

Reply:
I did think about that, but then decided I don't like silently ignoring the write. And I also didn't like adding a `Result` return type, with a read-only error. I like it explicit like this, where callers are responsible and we would notice when someone "misbehaves".